### PR TITLE
fix the output file name and diag time

### PR DESCRIPTION
**Pull request by junwang-noaa**
## Description

The output file name does not have correct time stamp when the output time is not a integer. E.g. output_fh=0.25, the file name is currently atmf00.nc and sfcf00.nc. Also for random output time, the diag_time may not be updated when the output time is not multiple of fhzero. This PR contains fixes for the two issues.

### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues should always be created before starting work on a PR branch!)
- fv3atm issue [#403](https://github.com/NOAA-EMC/fv3atm/issues/403)

## Testing

How were these changes tested?  Tested on hera. 
What compilers / HPCs was it tested with?  
Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)  
Have the ufs-weather-model regression test been run? On what platform?  
Yes on hera.
- Will the code updates change regression test baseline? If yes, why? Please show the baseline directory below.
- No

## Dependencies

If testing this branch requires non-default branches in other repositories, list them.
Those branches should have matching names (ideally)

ufs-weather-model PR
fv3atm PR

### CODE REVIEW

**Review by climbfuji — COMMENTED** [review]
I don't understand the logic, especially in lines 763 to 772. Is it possible to document that code? Just a brief one liner what each of those blocks does would be great.

**Comment by DusanJovic-NOAA** [comment]
Why do need all this logic with `lflname_fulltime`, why can we just switch to using the filename format that always include HH-MM-SS, which will simplify the code substantially. Scripting will be also simplified by always using the same file naming scheme, that does not change based on the value `output_fh`.

**Comment by junwang-noaa** [comment]
> 
> 
> Why do need all this logic with `lflname_fulltime`, why can we just switch to using the filename format that always include HH-MM-SS, which will simplify the code substantially. Scripting will be also simplified by always using the same file naming scheme, that does not change based on the value `output_fh`.

This is for backward compatible. Otherwise all the downstream jobs will be broken.

**Comment by DusanJovic-NOAA** [comment]
> > Why do need all this logic with `lflname_fulltime`, why can we just switch to using the filename format that always include HH-MM-SS, which will simplify the code substantially. Scripting will be also simplified by always using the same file naming scheme, that does not change based on the value `output_fh`.
> 
> This is for backward compatible. Otherwise all the downstream jobs will be broken.

Can't we announce the change and downstream jobs can update scripts. This is after all develop branch, we should be able to make breaking changes. Are we going to keep backward compatibility forever. Every change we make in any of the config file structure or output is potential breaking change for somebody.

File: fv3_cap.F90 [hunk 1]
```diff
@@ -746,6 +746,9 @@ subroutine InitializeAdvertise(gcomp, rc)
               endif
               do i=2,nfh
                 output_fh(i) = (i-1)*outputfh2(1) + output_startfh
+                ! Except fh000, which is the first time output, if any other of the 
+                ! output time is not integer hour, set lflname_fulltime to be ture, so the
```
Comment (climbfuji) [inline fv3_cap.F90]:
```suggestion
                ! output time is not integer hour, set lflname_fulltime to be true, so the 
```

File: fv3_cap.F90 [hunk 2]
```diff
@@ -760,6 +763,10 @@ subroutine InitializeAdvertise(gcomp, rc)
              count=noutput_fh, rc=rc)
           if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
           if( output_startfh == 0) then
+            ! if the output time in output_fh array contains first time stamp output,
```
Comment (climbfuji) [inline fv3_cap.F90]:
```suggestion
            ! If the output time in output_fh array contains first time stamp output,
```

File: fv3_cap.F90 [hunk 3]
```diff
@@ -760,6 +763,10 @@ subroutine InitializeAdvertise(gcomp, rc)
...
+            ! check the rest of output time, otherwise, check all the output time. 
+            ! if any of them is not integer hour, the history file names will
```
Comment (climbfuji) [inline fv3_cap.F90]:
```suggestion
            ! If any of them is not integer hour, the history file names will
```

File: fv3_cap.F90 [hunk 4]
```diff
@@ -773,6 +780,9 @@ subroutine InitializeAdvertise(gcomp, rc)
           else
             do i=1,noutput_fh
               output_fh(i) = output_startfh + output_fh(i)
+              ! when output_startfh >0, check all the output time, if any of
```
Comment (climbfuji) [inline fv3_cap.F90]:
```suggestion
              ! When output_startfh >0, check all the output time, if any of
```

File: fv3_cap.F90 [hunk 5]
```diff
@@ -773,6 +780,9 @@ subroutine InitializeAdvertise(gcomp, rc)
...
+              ! them is not integer hour, set lflname_fulltime to be ture. The
```
Comment (climbfuji) [inline fv3_cap.F90]:
```suggestion
              ! them is not integer hour, set lflname_fulltime to be true. The 
```

**Review by climbfuji — CHANGES_REQUESTED** [review]
Thanks for adding those comments. A few suggestions/corrections of typos, otherwise looks good.

**Comment by junwang-noaa** [comment]
@climbfuji I added some comments to explain, please check. Basically for frequency output, we will don't count the  the first time step output as it is labeled as fh000.

**Comment by climbfuji** [comment]
> @climbfuji I added some comments to explain, please check. Basically for frequency output, we will don't count the the first time step output as it is labeled as fh000.

I did, please look at/accept my suggestions.

**Comment by junwang-noaa** [comment]
@DusanJovic-NOAA Well, this will impact future operational implementations. As we don't know if we will have higher frequency output, I think we may need to get the downstream involved before we make the hard code change. Right now if the output time is non-integer (needs min/sec info), the output file will have full time stamp, otherwise if it's all integer hours, file names will only contain hours.

**Review by DusanJovic-NOAA — APPROVED** [review]

**Review by climbfuji — APPROVED** [review]

**Review by DeniseWorthen — APPROVED** [review]

**Pull request merged**